### ISSUE TITLE
DepressionFinderAndRouter hangs when nodata nodes set to inactive

### ISSUE THREAD

**Issue by AlexLipp** [landlab/landlab#922]
I'm attempting to perform a landscape evolution model on real topgraphic data but having issues with DepressionFinderAndRouter causing the script to hang when nodata values are set to inactive.

I'm loading in a topographic .asc file ([topo_SW_clip.zip](https://github.com/landlab/landlab/files/3141670/topo_SW_clip.zip)) from SRTM30 for the SW of England (attached) as a model grid. This is clipped so that areas which are in the sea are given -9999 values. This topography is then loaded into LandLab and I'm trying to apply the stream power law. However, noise in the topo has a series of depressions I want to fill in, but how I deal with the no-data values (-9999) can cause the DepressionFiller to seemingly crash.

## Problem Example

The following code hangs:
```
#Load in topo
mg,zr = read_esri_ascii("SW_preprocessing/topo_SW_clip.asc",name='topographic__elevation')
#Set ocean nodes to inactive
mg.set_nodata_nodes_to_inactive(zr, -9999)

#BCs
mg.status_at_node[mg.nodes_at_left_edge] = FIXED_VALUE_BOUNDARY
mg.status_at_node[mg.nodes_at_top_edge] = FIXED_VALUE_BOUNDARY
mg.status_at_node[mg.nodes_at_bottom_edge] = FIXED_VALUE_BOUNDARY
#Set eastern boundary to closed because of somerset (for now)
mg.status_at_node[mg.nodes_at_right_edge] = CLOSED_BOUNDARY
#Dirichlet boundary conditions
zr[mg.boundary_nodes] = 0

frr = FlowAccumulator(
    mg,
    'topographic__elevation',
    flow_director = 'FlowDirectorD8')
frr.run_one_step()

df = DepressionFinderAndRouter(mg)
df.map_depressions() #### <--- CRASH OCCURS HERE ####
imshow_grid(mg,'is_pit')
```
### Error Message
When I end the processing I get the following traceback: 
Traceback (most recent call last):

```
  File "<ipython-input-58-feba3a1b9c19>", line 18, in <module>
    df.map_depressions()

  File "~/anaconda3/lib/python3.7/site-packages/landlab/components/flow_routing/lake_mapper.py", line 939, in map_depressions
    self._identify_depressions_and_outlets(reroute_flow)

  File "~/anaconda3/lib/python3.7/site-packages/landlab/components/flow_routing/lake_mapper.py", line 848, in _identify_depressions_and_outlets
    self.find_depression_from_pit(pit_node, reroute_flow)

  File "~/anaconda3/lib/python3.7/site-packages/landlab/components/flow_routing/lake_mapper.py", line 791, in find_depression_from_pit
    nodes_this_depression

  File "~/anaconda3/lib/python3.7/site-packages/landlab/components/flow_routing/lake_mapper.py", line 433, in find_lowest_node_on_lake_perimeter
    or self.flood_status[nbr] == _FLOODED
```
The same thing happens if the nodata nodes are set to be closed via `mg.set_nodata_nodes_to_closed(zr, -9999)`.
I don't really want them closed because it's unrealistic but it's interesting that it gives the same results. Is there a similar function for setting no data nodes to be a `FIXED_VALUE_BOUNDARY`? 

## Hacky solution

However when the sea nodes are set to 0 (manually), the Depression Finder does successfully complete. This is a hacky solution and does produce meaningful results on the land and I can just end up clipping this in postprocessing. However, it does mean that rivers end up flowing over what in reality is the ocean, which isn't great. I don't like this solution because it's a bit of a hackjob, although it does produce the correct(ish) results. Ideally I want to model the nodes which correspond to no-data to either be inactive or set to be fixed value boundaries, but this causes the depression router to crash.

```
mg,zr = read_esri_ascii("SW_preprocessing/topo_SW_clip.asc",name='topographic__elevation')

sea_nodes = np.where(zr == -9999.0)
zr[sea_nodes] = 0

mg.status_at_node[mg.nodes_at_left_edge] = FIXED_VALUE_BOUNDARY
mg.status_at_node[mg.nodes_at_top_edge] = FIXED_VALUE_BOUNDARY
mg.status_at_node[mg.nodes_at_bottom_edge] = FIXED_VALUE_BOUNDARY
mg.status_at_node[mg.nodes_at_right_edge] = CLOSED_BOUNDARY
zr[mg.boundary_nodes] = 0

frr = FlowAccumulator(
    mg,
    'topographic__elevation',
    flow_director = 'FlowDirectorD8')
frr.run_one_step()

df = DepressionFinderAndRouter(mg)
df.map_depressions()
imshow_grid(mg,'is_pit')
```

## Questions 

So to summarise 
1) Why does the depression router hang when no data nodes are set to inactive nodes or closed? 

2) Is there an equivalent function to ` set_nodata_nodes_to_closed` but for `FIXED_VALUE_BOUNDARY`.

Thanks all! Apart from this, everything looks great.

**Comment by kbarnhart** [comment]
@AlexLipp Thanks for the question and for providing a thorough example! I'm assigning @gregtucker and @SiccarPoint as the primary code authors of the DepressionFinderAndRouter. 

I will note that the `set_nodata_nodes_to_inactive` is an old deprecated function (that has be superseded by `set_nodata_nodes_to_closed`). I would recommend not using it.

**Comment by AlexLipp** [comment]
Thanks! I won't use the depreciated function!

**Comment by gregtucker** [comment]
@AlexLipp did using the newer function `set_nodata_nodes_to_closed` solve the problem?

**Comment by AlexLipp** [comment]
@gregtucker I haven't used `set_nodata_nodes_to_closed` as it's not appropriate for my uses. I don't want to close the nodes, I want to set them to a fixed boundary. I'm trying to set it up as a coastline, so if the no-data nodes are set to closed, nothing will be able to flow into the sea which doesn't fit. Ideally there should be a  function which sets them to be `fixed-value` boundary. At present, I am manually setting all the no-data values to 0 which is in essence the same thing, but it does mean that I still have to run the flow-router/depression-router/fastscaper over the sea-nodes which won't change, which is a little inefficient. 

Is there a way that a `set_nodata_nodes_to_fixed` function could be implemented in the next update? Seems like a useful thing to implement for coastlines.

**Comment by SiccarPoint** [comment]
Would something like this work for your case?
```
>>> from landlab import FIXED_VALUE_BOUNDARY
>>> [grid instantiation & data setup stuff]
>>> grid.status_at_node[np.equal(
...         grid.at_node['topographic__elevation'], [your nodata val]
...     )] = FIXED_VALUE_BOUNDARY
```

The `status_at_node` property on the grid is extremely smart. If you manually update it like this, it will silently and automatically update all the other necessary grid boundary condition properties (e.g., link statuses) to reflect the changes you make to it "by hand" like this.

**Comment by gregtucker** [comment]
I agree with the fix @SiccarPoint suggestions: set the nodata nodes to fixed value "manually" (by modifying `status_at_node` for the appropriate nodes).

I'm guessing that the reason the DepressionFinderAndRouter hangs when nodata nodes are all set closed is that you have no open boundary (fixed value) nodes to act as an outlet.

Please let us know whether the above fix works.

**Comment by AlexLipp** [comment]
Hi all, 

Yes this has worked fine. Thank you! 
Closing this now.
